client/api/src/notifications.rs

```diff
@@ -21,30 +21,49 @@
 use std::{
 	collections::{HashMap, HashSet},
 	pin::Pin,
-	sync::{Arc, Weak},
+	sync::Arc,
 	task::Poll,
 };
 
-use fnv::{FnvHashMap, FnvHashSet};
 use futures::Stream;
-use parking_lot::Mutex;
-use prometheus_endpoint::{register, CounterVec, Opts, Registry, U64};
-use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
-use sp_core::{
-	hexdisplay::HexDisplay,
-	storage::{StorageData, StorageKey},
-};
+
+use prometheus_endpoint::Registry as PrometheusRegistry;
+
+use sc_utils::pubsub::{channels::TracingUnbounded, Hub, Receiver};
+use sp_core::storage::{StorageData, StorageKey};
 use sp_runtime::traits::Block as BlockT;
 
+mod registry;
+
+use registry::Registry;
+
+#[cfg(test)]
+mod tests;
+
+/// A type of a message delivered to the subscribers
+pub type StorageNotification<Hash> = (Hash, StorageChangeSet);
+
 /// Storage change set
 #[derive(Debug)]
 pub struct StorageChangeSet {
-	changes: Arc<Vec<(StorageKey, Option<StorageData>)>>,
-	child_changes: Arc<Vec<(StorageKey, Vec<(StorageKey, Option<StorageData>)>)>>,
+	changes: Arc<[(StorageKey, Option<StorageData>)]>,
+	child_changes: Arc<[(StorageKey, Vec<(StorageKey, Option<StorageData>)>)]>,
 	filter: Keys,
 	child_filters: ChildKeys,
 }
 
+/// Manages storage listeners.
+#[derive(Debug)]
+pub struct StorageNotifications<Block: BlockT>(
+	Hub<TracingUnbounded<StorageNotification<Block::Hash>>, Registry>,
+);
+
+/// Type that implements `futures::Stream` of storage change events.
+pub struct StorageEventStream<H>(Receiver<TracingUnbounded<StorageNotification<H>>, Registry>);
```

Comment:
Can this not just be a type def?

Reply:
In order to define it as just a type-def I would need to make the `Registry` visible from outside.

The `Registry` has no other need to be exposed.